ui/shared/status/StatusChatInput.qml

```diff
@@ -0,0 +1,557 @@
+
+        let textBeforeCursor = Emoji.deparseFromParse(plainText.substr(0, i));
+        return {
+            cursor: countEmojiLengths(plainText.substr(0, i)) + messageInputField.cursorPosition,
+            data: Emoji.deparseFromParse(textBeforeCursor),
+        };
+    }
+
+    function emojiHandler(event) {
+        let message = extrapolateCursorPosition();
+        pollEmojiEvent(message);
+
+        // state machine to handle different forms of the emoji event state
+        if (!emojiEvent && isColonPressed) {
+            return (message.data.length <= 1 || isSpace(message.data.charAt(message.cursor - 1))) ? true : false;
+        } else if (emojiEvent && isColonPressed) {
+            const index = message.data.lastIndexOf(':', message.cursor - 2);
+            if (index >= 0 && message.cursor > 0) {
+                const shortname = message.data.substr(index, message.cursor);
+                const codePoint = Emoji.getEmojiUnicode(shortname);
+                if (codePoint !== undefined) {
+                    const newMessage = message.data
+                        .replace(shortname, Emoji.fromCodePoint(codePoint))
+                        .replace(/ /g, "&nbsp;");
+                    messageInputField.remove(0, messageInputField.cursorPosition);
+                    insertInTextInput(0, Emoji.parse(newMessage, '26x26'));
+                }
+                return false;
+            }
+            return true;
+        } else if (emojiEvent && isKeyValid(event.key) && !isColonPressed) {
+            // popup
+            return true;
+        } else if (emojiEvent && !isKeyValid(event.key) && !isColonPressed) {
+            return false;
+        }
+        return false;
+    }
+
+    function countEmojiLengths(value) {
+        const match = Emoji.getEmojis(value);
+        var length = 0;
+
+        if (match && match.length > 0) {
+            for (var i = 0; i < match.length; i++) {
+                length += Emoji.deparseFromParse(match[i]).length;
+            }
+            length = length - match.length;
+        }
+        return length;
+    }
+
+    // check if user has placed cursor near valid emoji colon token
+    function pollEmojiEvent(message) {
+        const index = message.data.lastIndexOf(':', message.cursor);
+        if (index >= 0) {
+            emojiEvent = validSubstr(message.data.substr(index, message.cursor - index));
+        } 
+    }
+
+    function validSubstr(substr) {
+        for(var i = 0; i < substr.length; i++) {
+            var c = substr.charAt(i);
+            if (isSpace(c) === true || isPunct(c) === true)
+                return false;
+        }
+        return true;
+    }
+
+    function isSpace(c) {
+        if (/( |\t|\n|\r)/.test(c))
+            return true;
+        return false;
+    }
+
+    function isPunct(c) {
+        if (/(!|\@|#|\$|%|\^|&|\*|\(|\)|_|\+|\||-|=|\\|{|}|[|]|"|;|'|<|>|\?|,|\.|\/)/.test(c))
+            return true;
+        return false;
+    }
```

Comment:
I moved those two functions to Utils. It's possible that you're not rebased on latest master. Hopefully we don't lose some of the big fixes and features I added

Reply:
You probably can remove those two functions